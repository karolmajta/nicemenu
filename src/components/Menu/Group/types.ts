import { IconName } from "../Item/Icon/types";
import React from "react";
import { AllowedShortcutKey, ItemProps } from "../Item/types";

export interface GroupItemProps {
  icon: IconName;
  label: string;
  shortcut?: Array<Array<AllowedShortcutKey>>;
  goto?: string;
  onSelect?: () => void;
  listIndex: number;
}

export interface GroupProps {
  label: string;
  items: Array<GroupItemProps>;
  activeListIndex: number | null;
  className?: string;
  classNames?: Array<string>;
  components?: {
    Item?: React.ComponentType<ItemProps>;
  };
  onItemMouseMove?: (index: number) => void;
  onGoto: (gotoPath: string) => void;
}
