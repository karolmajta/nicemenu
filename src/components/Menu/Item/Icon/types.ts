export type IconName =
  | "bolt"
  | "dot"
  | "inbox"
  | "loader"
  | "pencil"
  | "reschedule";

export interface IconProps {
  name: IconName;
  className?: string;
  classNames?: Array<string>;
}
