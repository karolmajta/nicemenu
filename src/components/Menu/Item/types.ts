import React from "react";
import { IconName, IconProps } from "./Icon/types";

export type AllowedShortcutKey =
  | "a"
  | "b"
  | "c"
  | "d"
  | "e"
  | "f"
  | "g"
  | "h"
  | "i"
  | "j"
  | "k"
  | "l"
  | "m"
  | "n"
  | "o"
  | "p"
  | "q"
  | "r"
  | "s"
  | "t"
  | "u"
  | "v"
  | "w"
  | "x"
  | "y"
  | "z"
  | "Shift";

export interface ItemProps {
  label: string;
  icon?: IconName;
  shortcut?: Array<Array<AllowedShortcutKey>>;
  listIndex: number;
  highlighted: boolean;
  goto?: string;
  className?: string;
  classNames?: Array<string>;
  components?: {
    Icon?: React.ComponentType<IconProps>;
  };
  onMouseMove?: (listIndex: number) => void;
  onSelect?: () => void;
  onGoto: (menuKey: string) => void;
}
