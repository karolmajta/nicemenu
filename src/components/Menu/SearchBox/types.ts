export interface SearchBoxProps {
  placeholder: string;
  searchTerm: string;
  focused: boolean;
  className?: string;
  classNames?: Array<string>;
  onArrowUp: () => void;
  onArrowDown: () => void;
  onChange: (searchTerm: string) => void;
  onFocus: () => void;
}
