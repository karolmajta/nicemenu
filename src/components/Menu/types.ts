import { IconName } from "./Item/Icon/types";
import React from "react";
import { ItemProps, AllowedShortcutKey } from "./Item/types";
import { GroupProps } from "./Group/types";
import { SearchBoxProps } from "./SearchBox/types";

export interface MenuItem {
  icon: IconName;
  label: string;
  shortcut?: Array<Array<AllowedShortcutKey>>;
  goto?: string;
  onSelect?: () => void;
}

export interface MenuGroup {
  label: string;
  items: Array<MenuItem>;
}

export interface MenuProps {
  title: string;
  items: Array<MenuItem>;
  groups: Array<MenuGroup>;
  submenus: {
    [submenuName: string]: MenuProps;
  };
  className?: string;
  classNames?: Array<string>;
  components?: {
    Item?: React.ComponentType<ItemProps>;
    Group?: React.ComponentType<GroupProps>;
    SearchBox?: React.ComponentType<SearchBoxProps>;
  };
}
