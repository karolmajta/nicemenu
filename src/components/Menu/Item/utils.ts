export const KEY_SEPARATOR = "+";

export const IGNORED_KEYS = new Set([
  "ArrowUp",
  "ArrowDown",
  "ArrowLeft",
  "ArrowRight",
  "Enter",
  "Escape",
]);
