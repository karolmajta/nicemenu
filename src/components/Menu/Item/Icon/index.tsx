import React from "react";
import cx from "classnames";
import styled from "styled-components";

import { IconProps } from "./types";
import { iconMapping } from "./utils";

export function IconUnstyled({ name, className, classNames }: IconProps) {
  return (
    <div
      data-testid="Laserfocus-Menu-Item-Icon"
      className={cx(
        "Laserfocus-Menu-Item-Icon",
        className,
        ...(classNames || [])
      )}
    >
      <img src={iconMapping[name]} alt={name} />
    </div>
  );
}

const Icon = styled(IconUnstyled)`
  display: flex;
  justify-content: center;
  align-items: center;
  margin-left: 20px;
  margin-right: 10px;

  img {
    display: block;
    width: 16px;
    height: 16px;
  }
`;

export default Icon;
