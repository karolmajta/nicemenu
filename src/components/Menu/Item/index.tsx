import React, { useEffect, useState, useContext } from "react";
import cx from "classnames";
import styled from "styled-components";

import WindowContext from "../../../contexts/WindowContext";
import { ItemProps } from "./types";
import { IGNORED_KEYS, KEY_SEPARATOR } from "./utils";
import Icon, { IconUnstyled } from "./Icon";

export function ItemUnstyled({
  label,
  icon,
  shortcut,
  className,
  classNames,
  components,
  listIndex,
  highlighted,
  goto,
  onMouseMove,
  onSelect,
  onGoto,
}: ItemProps) {
  const [currentlyPressedKeys, setCurrentlyPressedKeys] = useState<Set<string>>(
    new Set()
  );
  const window = useContext(WindowContext);

  const [shortcutIndex, setShortcutIndex] = useState(0);

  const IconComponent = components?.Icon || IconUnstyled;

  useEffect(() => {
    const onKeyUp = (e: KeyboardEvent) => {
      if (highlighted && e.key === "Enter" && onSelect) {
        onSelect();
      }
      if (highlighted && e.key === "Enter" && goto) {
        goto && onGoto(goto);
      }
    };
    window.addEventListener("keyup", onKeyUp);

    return () => {
      window.removeEventListener("keyup", onKeyUp);
    };
  }, [highlighted, onSelect, goto, onGoto, window]);

  useEffect(() => {
    const onKeyDown = (e: KeyboardEvent) => {
      if (
        !shortcut ||
        currentlyPressedKeys.has(e.key) ||
        IGNORED_KEYS.has(e.key)
      ) {
        setCurrentlyPressedKeys(new Set());
        return;
      }
      const newKeys = new Set([
        ...Array.from(currentlyPressedKeys),
        e.key.toLowerCase(),
      ]);
      const expectedKeys = shortcut[shortcutIndex];
      setCurrentlyPressedKeys(newKeys);
      if (
        newKeys.size > currentlyPressedKeys.size &&
        expectedKeys.every((key) => newKeys.has(key.toLowerCase()))
      ) {
        const expectedTogetherNow = Array.from(
          shortcut[shortcutIndex].map((k) => k.toLowerCase())
        )
          .sort()
          .join(KEY_SEPARATOR);
        const presentTogetherNow = Array.from(newKeys)
          .sort()
          .join(KEY_SEPARATOR);
        if (shortcutIndex + 1 === shortcut.length) {
          setShortcutIndex(0);
          if (expectedTogetherNow === presentTogetherNow) {
            onSelect && onSelect();
            goto && onGoto(goto);
          }
        } else {
          if (expectedTogetherNow === presentTogetherNow) {
            setShortcutIndex((i) => i + 1);
          } else {
            setShortcutIndex(0);
          }
        }
      }
    };

    const onKeyUp = (e: KeyboardEvent) => {
      if (!shortcut || IGNORED_KEYS.has(e.key)) {
        setCurrentlyPressedKeys(new Set());
        return;
      }
      setCurrentlyPressedKeys(
        (keys) =>
          new Set(Array.from(keys).filter((key) => key !== e.key.toLowerCase()))
      );
      let nonResetKeysArray: Array<string> = [];
      shortcut.slice(0, shortcutIndex).forEach((keys) => {
        nonResetKeysArray = [
          ...nonResetKeysArray,
          ...keys.map((k) => k.toLowerCase()),
        ];
      });
      const nonResetKeys = new Set(nonResetKeysArray);
      if (!nonResetKeys.has(e.key.toLowerCase())) {
        setShortcutIndex(0);
      }
    };

    window.addEventListener("keydown", onKeyDown);
    window.addEventListener("keyup", onKeyUp);

    return () => {
      window.removeEventListener("keydown", onKeyDown);
      window.removeEventListener("keyup", onKeyUp);
    };
  }, [
    shortcut,
    shortcutIndex,
    setShortcutIndex,
    currentlyPressedKeys,
    setCurrentlyPressedKeys,
    onSelect,
    goto,
    onGoto,
    window,
  ]);

  return (
    <div
      data-testid="Laserfocus-Menu-Item"
      onMouseMove={() => onMouseMove && onMouseMove(listIndex)}
      onClick={(e) => {
        onSelect && onSelect();
        goto && onGoto(goto);
        e.stopPropagation();
      }}
      className={cx(
        "Laserfocus-Menu-Item",
        { highlighted },
        className,
        ...(classNames || [])
      )}
    >
      {icon ? (
        <IconComponent name={icon} />
      ) : (
        <div className="icon-placeholder" />
      )}
      <div className="item-label">{label}</div>
      {shortcut && (
        <div className="item-shortcut">
          {shortcut.map((steps, i, shortcut) => (
            <div key={i} className="item-shortcut-parts">
              {steps.map((step, j) => (
                <div key={j} className="item-shortcut-step">
                  {step}
                </div>
              ))}
              {i < shortcut.length - 1 && (
                <div className="item-shortcut-part-separator">then</div>
              )}
            </div>
          ))}
        </div>
      )}
    </div>
  );
}

const Item = styled(ItemUnstyled).attrs((props) => ({
  ...props,
  components: {
    Icon: Icon,
  },
}))`
  display: flex;
  flex-direction: row;
  justify-content: flex-start;
  align-items: center;

  width: 100%;
  height: 44px;

  cursor: pointer;

  &.highlighted {
    background: rgba(228, 228, 238, 0.05);
  }

  .icon-placeholder {
    margin-left: 46px;
  }

  .item-label {
    display: flex;
    flex-direction: row;
    justify-content: flex-start;
    align-items: center;

    width: 100%;
    height: 36px;

    font-size: 15px;
    color: rgba(228, 228, 238, 0.8);
  }

  .item-shortcut {
    display: flex;
    flex-direction: row;
    justify-self: flex-end;

    margin-right: 20px;

    .item-shortcut-parts {
      display: flex;
      flex-direction: row;

      .item-shortcut-step {
        display: block;
        padding: 0 5px;

        height: 20px;
        min-width: 10px;
        margin-left: 12px;

        text-align: center;
        font-weight: 500;
        font-size: 12px;
        line-height: 20px;
        color: #828282;

        background: rgba(228, 228, 238, 0.1);
        border-radius: 4px;
      }

      .item-shortcut-part-separator {
        display: block;

        height: 20px;
        margin-left: 12px;

        font-weight: 500;
        font-size: 12px;
        line-height: 20px;
        color: #828282;
      }
    }
  }
`;

export default Item;
