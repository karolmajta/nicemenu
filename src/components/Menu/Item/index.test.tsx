import React from "react";
import { render, cleanup, fireEvent } from "@testing-library/react";
import Item from ".";
import { ItemProps } from "./types";

const renderItem = (props: Partial<ItemProps>) =>
  render(
    <Item
      label={props.label || "Example item"}
      icon={props.icon || "bolt"}
      shortcut={props.shortcut || [["a"], ["b", "c"], ["d"]]}
      listIndex={props.listIndex || 0}
      highlighted={props.highlighted || false}
      goto={props.goto || undefined}
      onMouseMove={props.onMouseMove || jest.fn()}
      onGoto={props.onGoto || jest.fn()}
      onSelect={props.onSelect || jest.fn()}
    />
  );

afterEach(() => {
  cleanup();
});

test("renders correctly", () => {
  const { getByTestId } = renderItem({});
  const itemDiv = getByTestId("Laserfocus-Menu-Item");
  expect(itemDiv).toMatchSnapshot();
});

test("calls onSelect when shortcut combination is pressed", () => {
  const onSelect = jest.fn();
  renderItem({ onSelect });
  fireEvent.keyDown(window, { key: "a" });
  fireEvent.keyUp(window, { key: "a" });
  fireEvent.keyDown(window, { key: "b" });
  fireEvent.keyDown(window, { key: "c" });
  fireEvent.keyUp(window, { key: "b" });
  fireEvent.keyUp(window, { key: "c" });
  fireEvent.keyDown(window, { key: "d" });
  expect(onSelect).toHaveBeenCalled();
});

test("calls onGoto when shortcut combination is pressed and goto is defined", () => {
  const onGoto = jest.fn();
  renderItem({ onGoto, goto: "some-menu" });
  fireEvent.keyDown(window, { key: "a" });
  fireEvent.keyUp(window, { key: "a" });
  fireEvent.keyDown(window, { key: "b" });
  fireEvent.keyDown(window, { key: "c" });
  fireEvent.keyUp(window, { key: "b" });
  fireEvent.keyUp(window, { key: "c" });
  fireEvent.keyDown(window, { key: "d" });
  expect(onGoto).toHaveBeenCalled();
});

test("calls onSelect when clicked", () => {
  const onSelect = jest.fn();
  const { getByTestId } = renderItem({ onSelect });
  const itemDiv = getByTestId("Laserfocus-Menu-Item");
  fireEvent.click(itemDiv);
  expect(onSelect).toHaveBeenCalled();
});

test("calls onGoto when clicked and goto is defined", () => {
  const onGoto = jest.fn();
  const { getByTestId } = renderItem({ onGoto, goto: "some-menu" });
  const itemDiv = getByTestId("Laserfocus-Menu-Item");
  fireEvent.click(itemDiv);
  expect(onGoto).toHaveBeenCalled();
});

test("calls onSelect when highlighted and enter is pressed", () => {
  const onSelect = jest.fn();
  renderItem({ onSelect, highlighted: true });
  fireEvent.keyUp(window, { key: "Enter" });
  expect(onSelect).toHaveBeenCalled();
});

test("calls onGoto when highlighted and enter is pressed and goto is defined", () => {
  const onGoto = jest.fn();
  renderItem({ onGoto, goto: "some-menu", highlighted: true });
  fireEvent.keyUp(window, { key: "Enter" });
  expect(onGoto).toHaveBeenCalled();
});
