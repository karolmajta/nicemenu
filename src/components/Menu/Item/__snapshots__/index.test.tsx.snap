// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  class="Laserfocus-Menu-Item sc-gtssRu iLlUnN"
  data-testid="Laserfocus-Menu-Item"
>
  <div
    class="Laserfocus-Menu-Item-Icon sc-bdnylx uunbO"
    data-testid="Laserfocus-Menu-Item-Icon"
  >
    <img
      alt="bolt"
      src="bolt.svg"
    />
  </div>
  <div
    class="item-label"
  >
    Example item
  </div>
  <div
    class="item-shortcut"
  >
    <div
      class="item-shortcut-parts"
    >
      <div
        class="item-shortcut-step"
      >
        a
      </div>
      <div
        class="item-shortcut-part-separator"
      >
        then
      </div>
    </div>
    <div
      class="item-shortcut-parts"
    >
      <div
        class="item-shortcut-step"
      >
        b
      </div>
      <div
        class="item-shortcut-step"
      >
        c
      </div>
      <div
        class="item-shortcut-part-separator"
      >
        then
      </div>
    </div>
    <div
      class="item-shortcut-parts"
    >
      <div
        class="item-shortcut-step"
      >
        d
      </div>
    </div>
  </div>
</div>
`;
